tests/Module/RateCalculatorTest.php

```diff
@@ -0,0 +1,142 @@
+				'cartWeightKg'      => 15,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 22,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 20,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 33,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => null,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 31,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 111,
+				CarrierOptionsDummyFactory::getProductValuePricingCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 1,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 222,
+				CarrierOptionsDummyFactory::getProductValuePricingCarrier(),
+				'totalProductValue' => 150,
+				'cartWeightKg'      => 15,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 222,
+				CarrierOptionsDummyFactory::getProductValuePricingCarrier(),
+				'totalProductValue' => 200,
+				'cartWeightKg'      => 20,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => 333,
+				CarrierOptionsDummyFactory::getProductValuePricingCarrier(),
+				'totalProductValue' => 300,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			[
+				'expectedCost'      => null,
+				CarrierOptionsDummyFactory::getProductValuePricingCarrier(),
+				'totalProductValue' => 400,
+				'cartWeightKg'      => 31,
+				'isCouponApplied'   => false,
+			],
+			'free shipping threshold must be reached'     => [
+				'expectedCost'      => 0,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 15000,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			'free shipping threshold must not be reached' => [
+				'expectedCost'      => 33,
+				CarrierOptionsDummyFactory::getNoFreeShippingLimitCarrier(),
+				'totalProductValue' => 15000,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => false,
+			],
+			'free shipping coupon must affect cost'       => [
+				'expectedCost'      => 0,
+				CarrierOptionsDummyFactory::getDefaultCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => true,
+			],
+			'free shipping coupon must not affect cost'   => [
+				'expectedCost'      => 33,
+				CarrierOptionsDummyFactory::getNoCouponCarrier(),
+				'totalProductValue' => 100,
+				'cartWeightKg'      => 25,
+				'isCouponApplied'   => true,
+			],
+		];
+	}
+
+	/**
+	 * @dataProvider calculationDataProvider
+	 */
+	public function testGetShippingRateCost( ?float $expectedCost, Options $carrierOptions, float $totalProductValue, float $cartWeight, bool $isCouponApplied ): void {
```

Comment:
the parameter names of this method should match the dataProvider keys.
See `$cartWeight` vs `cartWeightKg`